packages/@react-stately/color/src/useColorFieldState.ts

```diff
@@ -66,59 +72,107 @@ export function useColorFieldState(
+    let newColorValue = '';
+    if (colorValue) {
+      newColorValue = colorValue.toString('hex');
+    }
+    setInputValue(value === undefined ? '' : newColorValue);
+  };
+
+  let increment = () => {
+    safelySetColorValue((prevColor: Color) => {
+      let newValue = addColorValue(parsed.current, step);
+      // if we've arrived at the same value that was previously in the state, the
+      // input value should be updated to match
+      // ex type 4, press increment, highlight the number in the input, type 4 again, press increment
+      // you'd be at 5, then incrementing to 5 again, so no re-render would happen and 4 would be left in the input
+      if (newValue === prevColor) {
+        setInputValue(newValue.toString('hex'));
       }
-      return inputValue;
+      return newValue;
     });
-  }, [inputValue, colorValue, setInputValue]);
-
-  let increment = () => setColorValue((prevColor: Color) => addColorValue(prevColor, step));
-  let decrement = () => setColorValue((prevColor: Color) => addColorValue(prevColor, -step));
-  let incrementToMax = () => setColorValue((prevColor: Color) => addColorValue(prevColor, MAX_COLOR_INT));
-  let decrementToMin = () => setColorValue((prevColor: Color) => addColorValue(prevColor, -MAX_COLOR_INT));
-
-  let setFieldInputValue = (value: string) => {
-    value = value.match(/^#?[0-9a-f]{0,6}$/i)?.[0];
-    if (value !== undefined) {
-      if (!value.length && colorValue) {
-        setColorValue(null);
-        return;
-      }
-      try {
-        let newColor = parseColor(value.startsWith('#') ? value : `#${value}`);
-        setColorValue((prevColor: Color) => {
-          setInputValue(value);
-          return prevColor && prevColor.toHexInt() === newColor.toHexInt() ? prevColor : newColor;
-        });
-      } catch (err) {
-        setInputValue(value);
+  };
+  let decrement = () => {
+    safelySetColorValue((prevColor: Color) => {
+      let newValue = addColorValue(parsed.current, -step);
+      // if we've arrived at the same value that was previously in the state, the
+      // input value should be updated to match
+      // ex type 4, press increment, highlight the number in the input, type 4 again, press increment
+      // you'd be at 5, then incrementing to 5 again, so no re-render would happen and 4 would be left in the input
+      if (newValue === prevColor) {
+        setInputValue(newValue.toString('hex'));
       }
-    }
+      return newValue;
+    });
   };
+  let incrementToMax = () => safelySetColorValue(MAX_COLOR);
+  let decrementToMin = () => safelySetColorValue(MIN_COLOR);
 
-  let commit = () => {
-    setInputValue(colorValue ? colorValue.toString('hex') : '');
-  };
+  let validate = (value: string) => value === '' || !!value.match(/^#?[0-9a-f]{0,6}$/i)?.[0];
 
   return {
+    validate,
     colorValue,
     inputValue,
-    setInputValue: setFieldInputValue,
+    setInputValue,
     commit,
     increment,
```

Comment:
Also wonder what of this can be shared between NumberField and color field. 🤔

Reply:
I think a fair amount, but given that it's not quite as 1:1 as the aria behavior, maybe we wait until there is a third component that needs this kind of state?